dmriprep/workflows/base.py

```diff
@@ -331,176 +437,620 @@ def init_single_subject_wf(
+            if fmap_type not in ('epi', 'phasediff', 'phase'):
+                continue
+
+            if fmap_type == 'phase':
+                fmap_type = 'phase1'
+
+            this_priority = FMAP_PRIORITY[fmap_type]
+            if this_priority < priority:
+                priority = this_priority
+                fmap_file = fmap[fmap_type]
+                fmap_pointers[fmap_file] = fmap
+
+        fmap_tuple = (ref_file, metadata['PhaseEncodingDirection'], fmap_file, fmap_type)
+        parsed_dwis.append(fmap_tuple)
+        rationale.append(' * %s (PEDir %s): %s (Type %s)' % fmap_tuple)
+
+    groups = []
+    if not combine_all_dwis:
+        rationale.append("\nEach DWI was assigned to its own group for preprocessing")
+        for ref_file, pe_dir, fmap_file, fmap_type in parsed_dwis:
+            groups.append({
+                'dwi_series': [ref_file],
+                'fieldmap_info': fmap_pointers[fmap_file],
+                'dwi_series_pedir': pe_dir})
+        LOGGER.info('\n'.join(rationale))
+        return groups
+
+    # group by warped space, defined by pe_dir and fmap
+    groups = defaultdict(list)
+    for ref_file, pe_dir, fmap, fmap_type in parsed_dwis:
+        groups[(pe_dir, fmap)].append(ref_file)
+
+    # Also group by PE dir
+    pedir_groups = defaultdict(list)
+    for warp_group, member_images in groups.items():
+        pe_dir, _ = warp_group
+        pedir_groups[pe_dir].extend(member_images)
+
+    if ignore_fieldmaps:
+        output_groups = []
+        rationale.append("\nFieldmaps were ignored and images were grouped by PE Direction")
+        for pe_dir, member_images in pedir_groups.items():
+            output_groups.append({
+                'dwi_series': member_images,
+                'fieldmap_info': {'suffix': None},
+                'dwi_series_pedir': pe_dir})
+        LOGGER.info('\n'.join(rationale))
+        return output_groups
+
+    # Print the group memberships
+    rationale.append("\nThe following warped spaces were defined to contain:")
+    for warp_group, member_images in groups.items():
+        rationale.append(' * PEDir: %s, fmap: %s ' % warp_group)
+        for member_image in member_images:
+            rationale.append('    -> ' + member_image)
+
+    # group by pe dir and fmap
+    complete_groups = []
+    for warp_group, member_images in groups.items():
+        pe_dir, fmap_file = warp_group
+        complete_groups.append({
+            'dwi_series': member_images,
+            'fieldmap_info': fmap_pointers[fmap_file],
+            'dwi_series_pedir': pe_dir})
+
+    if prefer_dedicated_fmaps:
+        rationale.append('\nStrictly using data from fmaps/ for SDC')
+        LOGGER.info('\n'.join(rationale))
+        return complete_groups
+
+    # Ignoring dedicated fieldmaps, define warped spaces by PEDir
+    match_header = 'No opportunities for bidirectional PEPolar DWI series were found.'
+    group_keys = set(list(pedir_groups.keys()))
```

Comment:
```suggestion
    group_keys = set(pedir_groups.keys())
```
set will consume the generator directly